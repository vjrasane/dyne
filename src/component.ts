import { types as T } from "dyne-commons";

interface IPrototype {
  prototype: any;
}

export default abstract class Component implements IPrototype {
  props: object;
  prototype: any;
  children: T.VirtualDom[];

  constructor(props: object, children: T.VirtualDom[]) {
    this.props = props;
    this.children = children;
  }

  abstract view(): T.VirtualDom | T.VirtualDom[];
}
