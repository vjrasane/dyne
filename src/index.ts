export { engine, Update, Init, Cmd, Sub } from "./engine";
import dom from "./dom";
export default dom;
