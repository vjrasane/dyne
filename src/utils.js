export const isFunction = obj =>
  !!(obj && obj.constructor && obj.call && obj.apply);

export const flatten = arr => arr.flat(Infinity);

export const isString = obj => typeof obj === "string" || obj instanceof String;
