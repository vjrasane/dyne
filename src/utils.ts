import { types } from "dyne-commons";

export const element = (
  context: string,
  props?: object,
  children?: types.DomElement[]
): types.DyneElement => ({
  type: context,
  props: props || {},
  children: children || []
});
