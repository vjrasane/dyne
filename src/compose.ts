import { Update, Updated } from "./engine/core/update";
import Return from "./engine/effects/return";
import { isFunction, exists, always } from "./utils";
import { OpticalUpdate, Lens, Optical } from "./optical";

/**
 * Combine optical updates into a single update function
 *
 * @param opticals
 */
export const combine = <M>(opticals: OpticalUpdate<M, any>[]) => (
  msg: object,
  model: M
): Updated<M> =>
  opticals.reduce(
    <C>(updated: Return<M>, optical: OpticalUpdate<M, C>) =>
      updated.bind((m: M) => {
        // get child model from parent model
        const child = optical.lens.get(m);
        // call update with child
        const ret: Updated<C> = optical.func(msg, child);
        // if update returned nothing, use previous model
        if (!exists(ret)) {
          return Update(m);
        }
        // ensure each update value is stored as update monad
        const up: Return<C> = Update.wrap(ret);
        // map monad back to parent model and return
        return up.map((c: C) => optical.lens.set(c, m));
      }),
    // wrap model in update monad
    Update(model)
  );

/**
 *
 * Get array of optical updates based on given object structure
 *
 * @param struct
 * @param parentLens
 */
const getOpticals = <M>(
  struct: object,
  parentLens?: Lens<M, any>
): OpticalUpdate<M, any>[] => {
  // Flattened array of optical updates
  const opticals: OpticalUpdate<M, any>[] = [];

  Object.entries(struct).forEach(([field, value]) => {
    // Lens for the currently processed field
    const lens: Lens<M, any> = parentLens
      ? // If parent lens is present, map it with field lens
        parentLens.map(Lens.field(field))
      : // Otherwise use only field lens
        Lens.field(field);
    // If value is a function, assume it is an instance of update
    if (isFunction(value)) {
      // Create optical update with the field lens
      const optical: OpticalUpdate<M, any> = Optical(lens, value);
      // Push to list of optical updates
      opticals.push(optical);
    } else if (typeof value === "object") {
      // Recursively create opticals for each object field
      const childOpticals: OpticalUpdate<M, any>[] = getOpticals(value, lens);
      // Push child optical updates to flattened array
      opticals.push(...childOpticals);
    } else {
      // For any other field, create optical update with constant return value
      const optical: OpticalUpdate<M, any> = Optical(lens, always(value));
      // Push to list of optical updates
      opticals.push(optical);
    }
  });
  return opticals;
};

const compose = <M>(struct: object): Update<M> => combine(getOpticals(struct));

export default compose;
