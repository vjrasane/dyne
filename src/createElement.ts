import { utils, types } from "dyne-commons";
import { Context, DynamicElement } from "./types";
import { element } from "./utils";

export default (
  context: Context,
  props?: object,
  ...children: types.VirtualDom[]
): types.VirtualDom => {
  /**
   * Flatten any inner arrays created from fragments or array literals
   */
  const flatChildren: types.DomElement[] = utils.flatten(children);
  /**
   * Ensure props are non-null
   */
  const propsObj = props || {};

  /**
   * Dynamic JSX components are passed as functions.
   */
  if (utils.isFunction(context)) {
    return (<DynamicElement>context)(propsObj, flatChildren);
  }

  /**
   *  Raw JSX elements have a string type, props and children.
   */
  if (typeof context === "string") {
    return element(context, propsObj, flatChildren);
  }

  /**
   * Static JSX components are passed as already created elements.
   * Any props or children are ignored.
   */
  if (utils.exists(context) && typeof context === "object") {
    return context;
  }

  throw new Error(
    `Invalid element '${context}', expected string, function or element`
  );
};
