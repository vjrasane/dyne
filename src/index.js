import { isFunction, flatten, element, isString } from "~/utils";

const dyne = () => {
  // inner scope here
  return {
    createElement: (context, props, ...children) => {
      /**
       * Flatten any inner arrays created from fragments or array literals
       */
      const flatChildren = flatten(children);

      /**
       * Dynamic JSX components are passed as functions.
       */
      if (isFunction(context)) {
        return context(props, flatChildren);
      }

      /**
       *  Raw JSX elements have a string type, props and children.
       */
      if (isString(context)) {
        return element(context, props, flatChildren);
      }

      /**
       * Static JSX components are passed as already created elements.
       * Any props or children are ignored.
       */
      if (typeof context === "object") {
        return context;
      }

      throw new Error(
        `Invalid element '${context}', expected string, function or element`
      );
    },
    /**
     * Fragments are simply array containers for their children
     */
    Fragment: (_, children) => flatten(children)
  };
};

export default dyne();
