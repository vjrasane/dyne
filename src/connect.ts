import { OpticalUpdate, Lens, Optical } from "./optical";
import { Update } from "./engine/core/update";
import { combine } from "./compose";

type Optic<P, C> = Lens<P, C> | string[] | string;

const toLens = <P, C>(optic: Optic<P, C>): Lens<P, C> => {
  if (typeof optic === "string") {
    return Lens.field(optic);
  } else if (Array.isArray(optic)) {
    return Lens.path(...optic);
  } else {
    return optic;
  }
};

const connector = <M>() => {
  const opticals: OpticalUpdate<M, any>[] = [];

  return {
    connect: <U>(optic: Optic<M, U>, inner: Update<U>): void => {
      opticals.push(Optical(toLens(optic), inner));
    },
    /**
     * Due to pass by reference and delayed evaluation,
     * any optical updates added by connect will be included
     * in the combined update function
     */
    update: combine(opticals)
  };
};

export default connector;
