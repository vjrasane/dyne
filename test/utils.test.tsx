import { element } from "../src/utils";

describe("utils", () => {
  describe("element", () => {
    it("creates minimal element", () => {
      const elem = element("div");
      expect(elem).toMatchSnapshot();
    });

    it("creates basic element", () => {
      const elem = element("div", { prop: "value" }, ["child"]);
      expect(elem).toMatchSnapshot();
    });
  });
});
