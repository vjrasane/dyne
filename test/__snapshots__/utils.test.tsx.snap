// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils element creates basic element 1`] = `
Object {
  "children": Array [
    "child",
  ],
  "props": Object {
    "prop": "value",
  },
  "type": "div",
}
`;

exports[`utils element creates minimal element 1`] = `
Object {
  "children": Array [],
  "props": Object {},
  "type": "div",
}
`;
